### ISSUE TITLE
Forming cluster over artery with lanes > 1 fails in some situations

### ISSUE THREAD

**Issue by jrudolph** [akka/akka#23328]
It seems joining just times out. It is still unclear if `inbound-lanes` or `outbound-lanes` are the problem.

Reported for Akka 2.5.2.

**Comment by ktoso** [comment]
We now merged the Partition hub so what we have on master is not quite the same as what was being run when this was reported... Just a note to remember.

**Comment by jrudolph** [comment]
Good point, it was reported for Akka 2.5.2.

**Comment by t3hnar** [comment]
will the upgrade to 2.5.3 help ?

**Comment by ktoso** [comment]
Without having reproduced I'm not sure we can claim so with enough confidence.
Also, the new thing I mention will only be released in 2.5.4.

Thanks for reporting we'll definitely keep an eye on it.

**Comment by ktoso** [comment]
It was not yet merged actually, but it's ready to go in, refs: https://github.com/akka/akka/pull/23102

**Comment by jrudolph** [comment]
@t3hnar, good point I forgot to mention that you could try updating to 2.5.3. I had a look through the commits between 2.5.2 and 2.5.3 which do not suggest that an update would help but you can try of course.

**Comment by t3hnar** [comment]
I will upgrade to 2.5.3, but also not found something related changed :)

**Comment by patriknw** [comment]
@t3hnar The improvements of inbound-lanes > 1 were released in 2.5.4. It's now using the new PartitionHub and multiple lanes are also used for ActorSelection.

This ticket doesn't contain much information. Could you provide logs (preferably debug level) of when it happened or describe a way to reproduce it?

**Comment by t3hnar** [comment]
@patriknw I will check against akka 2.5.4 as soon as possible and provide logs in case this is still an issue, but hope this is really fixed by your changes.

Btw, this is not something I was able to reproduce locally, only on high performance server.

**Comment by patriknw** [comment]
ok thanks

**Comment by guidomedina** [comment]
I have been using in/out bound lanes > 1 for a while with artery (I know it had no effect before 2.5.4) and I haven't had any issues with a 15 JVMs running as a cluster on 5 physical machines,

**Comment by t3hnar** [comment]
I have three nodes, no load, akka 2.5.4

```
seed-nodes = [
      "akka://coreservices@10.30.8.62:9196",
      "akka://coreservices@10.30.8.63:9196",
      "akka://coreservices@10.30.8.64:9196"
    ]
```

I had a proper launch with `outbound-lanes/inbound-lanes = 1/1`

With `8/8` I can see that node `10.30.8.63` produced number of logs like the one below and then just stopped spamming, however did not join the cluster.

```
2017-08-17 13:22:04,231 ERROR [][][][][][coreservices-akka.actor.default-dispatcher-4][] a.r.a.Association(akka://coreservices) - Outbound message stream to [akka://coreservices@10.30.8.64:9196] failed. Restarting it. Handshake with [akka://coreservices@10.30.8.64:9196] did not complete within 20000 ms
akka.remote.artery.OutboundHandshake$HandshakeTimeoutException: Handshake with [akka://coreservices@10.30.8.64:9196] did not complete within 20000 ms
2017-08-17 13:22:04,285 WARN  [][][][][][coreservices-akka.actor.default-dispatcher-4][] a.c.JoinSeedNodeProcess - Couldn't join seed nodes after [13] attempts, will try again. seed-nodes=[akka://coreservices@10.30.8.62:9196, akka://coreservices@10.30.8.64:9196]
2017-08-17 13:22:04,285 DEBUG [][][][][][coreservices-akka.actor.default-dispatcher-4][] a.r.a.Association(akka://coreservices) - Starting outbound message stream to [akka://coreservices@10.30.8.62:9196] with [8] lanes
2017-08-17 13:22:04,810 DEBUG [][][][][][coreservices-akka.actor.default-dispatcher-3][] a.r.a.Association(akka://coreservices) - Starting outbound message stream to [akka://coreservices@10.30.8.64:9196] with [8] lanes
```

Will continue testing with `1/8` & `8/1`

**Comment by t3hnar** [comment]
No issue with
```
outbound-lanes = 1
inbound-lanes = 8
```


However it is reproducible with
```
outbound-lanes = 1
inbound-lanes = 8
```
And in all last cases I have the first started node having described problem.

**Comment by patriknw** [comment]
Thanks for narrowing it down. It can be good to know that having `outbound-lanes` > 1 is less important than inbound lanes. Because there will anyway be a separate stream for each outgoing connection, so messages to different target nodes are anyway in parallel. We will of course investigate and fix it, though.

**Comment by t3hnar** [comment]
I did one more experiment - deployed with `outbound-lanes = 3` and it does work :)
I can assume that chance to reproduce depends on number of `outbound-lanes`

**Comment by t3hnar** [comment]
I can reproduce the issue on my local machine with `outbound-lanes = 32`

**Comment by patriknw** [comment]
I tried to reproduce, but no success so far. Tried with for example `outbound-lanes = 32`.
Could you run with additional logging and share logs?
```
akka {
  loglevel = DEBUG
  log-config-on-start = on
  remote.log-sent-messages = on
  remote.log-received-messages = on
}
```

**Comment by t3hnar** [comment]
strange... cannot reproduce it anymore locally
Will have access window to preprod server next week to grab `log-sent-messages` and `log-received-messages` logs.

Btw we have more setting I did not mentioned before.
```
maximum-frame-size = 3megabytes
maximum-large-frame-size = 3megabytes
```

**Comment by patriknw** [comment]
That would be good, thanks. The test you are running is simply that the joining of the cluster fails, right?

**Comment by t3hnar** [comment]
@patriknw I'm just starting the app that has cluster, sharding, pub subbing, distributed data components.

**Comment by patriknw** [comment]
The reason for this can be the same as I have fixed in https://github.com/akka/akka/pull/23842
Sometimes the first message(s) could be lost, but that was due to the inbound lanes, not the outbound lanes. Btw, if you are using `joinSeedNodes` instead of `join` it will retry automatically and then it should work even if some message is lost.

Anyway, there are several improvements to the lanes in progress and I hope to have them all ready for upcoming 2.5.7. Therefore I suggest that we close this when those PRs have been merged.
